controllers/front/ProductController.php

```diff
@@ -1323,4 +1312,65 @@ protected function isValidCombination($productAttributeId, $productId)
 
         return false;
     }
+
+    /**
+     * From the current request, tries to determine the current category being browsed
+     *
+     * @return int|null
+     */
+    protected function guessCategoryIdFromRequest()
+    {
+        // Assure us the previous page was one of the shop
+        if (!isset($_SERVER['HTTP_REFERER']) || $_SERVER['HTTP_REFERER'] !== Tools::secureReferrer($_SERVER['HTTP_REFERER'])) {
+            return null;
+        }
+
+        // Check if the previous page was a category page
+        if (preg_match('~^.*(?<!\/content)\/([0-9]+)\-(.*[^\.])|(.*)id_(category|product)=([0-9]+)(.*)$~', $_SERVER['HTTP_REFERER'], $regs) !== 1) {
+            return null;
+        }
+
+        // If the previous page was a category and is a parent category of the product use this category as parent category
+        $id_object = false;
+        if (isset($regs[1]) && is_numeric($regs[1])) {
+            $id_object = (int) $regs[1];
+        } elseif (isset($regs[5]) && is_numeric($regs[5])) {
+            $id_object = (int) $regs[5];
+        }
+
+        if (!$id_object) {
+            return null;
+        }
+
+        foreach ([$_SERVER['HTTP_REFERER'], urldecode($_SERVER['HTTP_REFERER'])] as $previousPageLink) {
+            $id_category = $this->extractObjectCategoryFromLink($id_object, $previousPageLink);
+            if ($id_category !== null) {
+                return $id_category;
```

Comment:
```suggestion
                return (int) $id_category;
```

Reply:
No harm but seems a bit unnecessary if the `extractObjectCategoryFromLink` is typed hinted no ?

Reply:
@jbenezech Correct. I close this one. (Thanks ;) )